app/models/spree/gateway/stripe_connect.rb

```diff
@@ -0,0 +1,152 @@
+                                             gateway_customer_profile_id: response.params['id'],
+                                             gateway_payment_profile_id: response.params['default_source'] || response.params['default_card'])
+        else
+          payment.send(:gateway_error, response.message)
+        end
+      end
+
+      private
+
+      # In this gateway, what we call 'secret_key' is the 'login'
+      def options
+        options = super
+        options.merge(:login => Stripe.api_key)
+      end
+
+      def options_for_purchase_or_auth(money, creditcard, gateway_options)
+        options = {}
+        options[:description] = "Spree Order ID: #{gateway_options[:order_id]}"
+        options[:currency] = gateway_options[:currency]
+        options[:stripe_account] = stripe_account_id
+
+        creditcard = token_from_card_profile_ids(creditcard)
+
+        [money, creditcard, options]
+      end
+
+      def address_for(payment)
+        {}.tap do |options|
+          if address = payment.order.bill_address
+            options[:address] = {
+              address1: address.address1,
+              address2: address.address2,
+              city: address.city,
+              zip: address.zipcode
+            }
+
+            if country = address.country
+              options[:address][:country] = country.name
+            end
+
+            if state = address.state
+              options[:address].merge!(state: state.name)
+            end
+          end
+        end
+      end
+
+      def update_source!(source)
+        source.cc_type = CARD_TYPE_MAPPING[source.cc_type] if CARD_TYPE_MAPPING.include?(source.cc_type)
+        source
+      end
+
+      def card_to_store(source)
+        source = update_source!(source)
+        if source.number.blank? && source.gateway_payment_profile_id.present?
+          # StripeJS Token
+          source.gateway_payment_profile_id
+        else
+          # Spree::CreditCard object
+          source
+        end
+      end
+
+      def token_from_card_profile_ids(creditcard)
+        token_or_card_id = creditcard.gateway_payment_profile_id
+        customer = creditcard.gateway_customer_profile_id
+
+        return nil if token_or_card_id.blank?
+
+        # Assume the gateway_payment_profile_id is a token generated by StripeJS
+        return token_or_card_id if customer.blank?
+
+        # Assume the gateway_payment_profile_id is a Stripe card_id
+        # So generate a new token, using the customer_id and card_id
+        tokenize_instance_customer_card(customer, token_or_card_id)
+      end
+
+      def tokenize_instance_customer_card(customer, card)
+        token = Stripe::Token.create({card: card, customer: customer}, stripe_account: stripe_account_id)
+        token.id
+      rescue Stripe::StripeError => e
+        Rails.logger.error("Stripe Error: #{e}")
+        nil
```

Comment:
Shouldn't we raise when we couldn't create the token? IMO we better show the failure and fix it than finding incosistent data leading to bugs later on.

Reply:
TODO

Reply:
What's the plan on this @oeoeaio ?

Reply:
@enricostano Resolved